### PR TITLE
feat(cardano): add support for babbage features

### PR DESCRIPTION
**Pull request by davidmisiak**
Changes:
- new Babbage transaction elements serialization and  the corresponding logic and UI changes
  - collateral return - an output which can be added to a transaction to contain unspent collateral. It would only be considered a real tx output if Plutus script validation fails and collateral would be spent. Useful so that any input can be used as collateral.
  - total collateral - if the total ADA amount in collateral inputs doesn't match total collateral the transaction is invalid.
  - reference inputs - inputs required for script validation.
  - output inline datum - the whole serialized datum can now be included in the output instead of just the datum hash. Is sent as chunks since it can be quite long. Only first chunk is shown to the user.
  - output reference script - used for script pre-publishing.  Is sent as chunks since it can be quite long. Only first chunk is shown to the user.
- "post-alonzo" output serialization as a map
  - Outputs can now be serialized in two formats. The legacy, array format or now also as a map. See [CDDL](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/babbage/test-suite/cddl-files/babbage.cddl#L75).
- required signers allowed in ORDINARY and MULTISIG txs - there should be no harm in allowing this and some dapps misuse this field fir whatever they need.
- datum hash allowed in non-script outputs - apparently this can be useful to "tag" inputs.

[Babbage CDDL](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/babbage/test-suite/cddl-files/babbage.cddl)

### CODE REVIEW

File: core/src/apps/cardano/sign_tx/signer.py [hunk 1]
```diff
@@ -458,6 +608,66 @@ def _validate_token(
...
+        self,
+        inline_datum_cbor: HashBuilderEmbeddedCBOR,
+        inline_datum_size: int,
+        should_show: bool,
+    ) -> None:
+        assert inline_datum_size > 0
+        self.has_hidden_data = True
+
+        chunks_count = self._get_chunks_count(inline_datum_size)
+        for i in range(chunks_count):
```
Comment (gabrielKerekes) [inline core/src/apps/cardano/sign_tx/signer.py:623]: nit:
```suggestion
        for chunk_number in range(chunks_count):
```

File: core/src/apps/cardano/sign_tx/ordinary_signer.py [hunk 2]
```diff
@@ -30,6 +30,9 @@ def _validate_tx_init(self) -> None:
         if (
```
Comment (gabrielKerekes) [inline core/src/apps/cardano/sign_tx/ordinary_signer.py:30]: Looking at these conditions now, I'd consider adding a `_assert_tx_init_cond` (or similar name) to validate tx init in the signers so that when reading the conditions it would be apparent what needs to be fulfilled. Currently when reading the code I need to always reverse the conditions in my head 🤷 E.g.:
```
self._assert_tx_init_cond(self.msg.collateral_inputs_count == 0)
self._assert_tx_init_cond(self.msg.required_signers_count == 0)
self._assert_tx_init_cond(self.msg.has_collateral_return == False)
self._assert_tx_init_cond(self.msg.total_collateral is None)
self._assert_tx_init_cond(self.msg.reference_inputs_count == 0)
```

In `PoolOwnerSigner` we could IMHO merge the two "condition groups" into `Invalid tx signing request`.

**Review by gabrielKerekes — COMMENTED** [review]
Just a preliminary review. I'll review once more when it's finalized. Overall it looks good, I'm just a little disturbed by the "almost" repeating code in some cases e.g. `_process_inline_datum` and `_process_reference_script`, `_show_output_init` and `_show_collateral_return_init`, but I wasn't able to think of a reasonable way to extract it - I suppose you tried too, we can discuss it together to see if we can't come up with something.

File: core/src/apps/cardano/sign_tx/signer.py [hunk 3]
```diff
@@ -379,6 +440,95 @@ def _is_change_output(self, output: messages.CardanoTxOutput) -> bool:
...
+        else:
+            # [amount, asset_groups] is added to the dict.
+            output_value_list: HashBuilderList = HashBuilderList(2)
+            with output_dict.add(POST_ALONZO_OUTPUT_KEY_AMOUNT, output_value_list):
+                await self._process_output_value(output_value_list, output, should_show)
+
+        if output.datum_hash is not None:
+            if should_show:
+                await layout.confirm_datum_hash(self.ctx, output.datum_hash)
+            output_dict.add(POST_ALONZO_OUTPUT_KEY_DATUM, output.datum_hash)
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/sign_tx/signer.py:487]: Right, thanks for noticing!

[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/sign_tx/signer.py:487]: Hm, but simply wappring inline datum like `[1, inline_datum_cbor]` won't work, because `inline_datum_cbor` is a hash builder, not a primitive type. I think we need a separate `HashBuilderList` for this.

File: core/src/apps/cardano/sign_tx/ordinary_signer.py (hunk 2 above)
Comment (davidmisiak) [reply core/src/apps/cardano/sign_tx/ordinary_signer.py:30]: I agree.

**Comment by davidmisiak** [comment]
Thanks. As discussed offline, I'm not perfectly satisfied with some of the changes either, let's see if we come up with some improvements. https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/a8db49efe3ea94f064e1e589361b77170fce3fe9 should fix the things you pointed out.

File: core/src/apps/cardano/sign_tx/pool_owner_signer.py [hunk 4]
```diff
@@ -56,7 +50,12 @@ async def _confirm_tx(self, tx_hash: bytes) -> None:
 
     def _validate_output(self, output: messages.CardanoTxOutput) -> None:
         super()._validate_output(output)
-        if output.address_parameters is not None or output.datum_hash is not None:
+        if (
+            output.address_parameters is not None
+            or output.datum_hash is not None
+            or output.inline_datum_size is not None
+            or output.reference_script_size is not None
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/sign_tx/pool_owner_signer.py:57]: Thanks, fixed (though I used `> 0` instead of `!= 0`, to be consistent with other similar conditions).

File: common/protob/messages-cardano.proto [hunk 5]
```diff
@@ -46,6 +46,11 @@ enum CardanoNativeScriptHashDisplayFormat {
     POLICY_ID = 2;
 }
 
+enum CardanoTxOutputSerializationFormat {
+    LEGACY = 0;
```
Comment (gabrielKerekes) [inline common/protob/messages-cardano.proto:50]: Just noting this so we don't forget to rename it according to the discussion https://github.com/vacuumlabs/cardano-hw-interop-lib/pull/7#discussion_r895865454. There's also a mention of post-alonzo in the README so that should perhaps also be udpated when this gets renamed.

File: core/src/apps/cardano/sign_tx/signer.py [hunk 6]
```diff
@@ -458,6 +616,62 @@ def _validate_token(
         if len(token.asset_name_bytes) > MAX_ASSET_NAME_LENGTH:
             raise INVALID_TOKEN_BUNDLE
 
+    # inline datum
+
+    async def _process_inline_datum(
```
Comment (gabrielKerekes) [inline core/src/apps/cardano/sign_tx/signer.py:621]: I tried to extract the chunk processing into a separate function in a separate branch https://github.com/vacuumlabs/trezor-firmware/commit/85f0e79c020bac7a3fff91a034fa840b2ecd5c6f. I'm 50/50 on whether it's better or not. What do you think?

**Review by gabrielKerekes — COMMENTED** [review]
I added a couple more suggestions/thoughts. I think there's nothing major.

File: common/protob/messages-cardano.proto [hunk 7]
```diff
@@ -210,6 +215,9 @@ message CardanoSignTxInit {
     optional bytes script_data_hash = 16;
     required uint32 collateral_inputs_count = 17;
     required uint32 required_signers_count = 18;
+    required bool has_collateral_return = 19;
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply common/protob/messages-cardano.proto:218]: Yeah well, we introduced the collateral input type so that it's easier to add new attributes in case the chain adds some (but it seems rather unlikely to me atm). I added the reference input type to be consistent with collateral inputs, but adding a new type for collateral return when it's just a more constrained output (and is likely to stay that way) seems too far-fetched to me :thinking: What do you think?

File: core/src/apps/cardano/README.md [hunk 8]
```diff
@@ -14,17 +14,17 @@ REVIEWER = Jan Matejek <jan.matejek@satoshilabs.com>, Tomas Susanka <tomas.susan
...
 
-[Plutus CDDL spec](https://github.com/input-output-hk/cardano-ledger/blob/9c3b4737b13b30f71529e76c5330f403165e28a6/eras/alonzo/test-suite/cddl-files/alonzo.cddl).
+[Delegation Design Spec](https://hydra.iohk.io/build/2006688/download/1/delegation_design_spec.pdf) - contains information about delegation (addresses, certificates, withdrawals, ...).
 
 [Byron address format](https://github.com/input-output-hk/cardano-wallet/wiki/About-Address-Format---Byron).
 
 [The Shelley 1852' purpose and staking path](https://github.com/input-output-hk/implementation-decisions/blob/e2d1bed5e617f0907bc5e12cf1c3f3302a4a7c42/text/1852-hd-chimeric.md).
 
-[cbor.me](http://cbor.me/) - very useful tool for CBOR inspection.
+[cbor.me](http://cbor.me/), [cbor.nemo157.com](https://cbor.nemo157.com/), [VS Code extension](https://marketplace.visualstudio.com/items?itemName=gabrielkerekes.cbor) - useful tools for CBOR inspection.
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/README.md:27]: :100:

File: common/tests/fixtures/cardano/sign_tx.json [hunk 9]
```diff
@@ -1819,6 +1995,242 @@
...
+            "type": 1,
+            "pub_key": "5d010cf16fdeff40955633d6c565f3844a288a24967cf6b76acbeb271b4f13c1",
+            "signature": "e9ecd38c4c8219eca08971a6949971f3afb33e64f727e9958bfc51b42b913f445adec954d938b9bad0759740979d45f5e97da1581d1d588ff0acfbf53232e90e",
+            "chain_code": null
+          }
+        ]
+      }
+    },
+    {
+      "description": "Ordinary transaction with inline datum, reference input and external address",
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply common/tests/fixtures/cardano/sign_tx.json]: https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/7b8229ab54c69b3af28d6c204c1a15db9ef0917a

File: tests/device_tests/cardano/test_sign_tx.py [hunk 10]
```diff
@@ -143,6 +162,9 @@ def test_cardano_sign_tx_failed(client: Client, parameters, result):
                 script_data_hash=script_data_hash,
                 collateral_inputs=collateral_inputs,
                 required_signers=required_signers,
+                collateral_return=collateral_return,
+                total_collateral=parameters.get("total_collateral"),
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply tests/device_tests/cardano/test_sign_tx.py]: https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/64eed36fc454b1098e16de8d055c92b387225abd

File: core/src/apps/cardano/README.md [hunk 11]
```diff
@@ -312,7 +322,7 @@ When one wants to spend funds from the Plutus script address (which is possible
 You can use a combination of [cardano-node](https://github.com/input-output-hk/cardano-node) and cardano-cli (part of the cardano-node repo) to submit a transaction.
 
 ## Serialization format
-Cardano uses [CBOR](https://www.rfc-editor.org/info/rfc7049) as a serialization format. [Here](https://github.com/input-output-hk/cardano-ledger/blob/9c3b4737b13b30f71529e76c5330f403165e28a6/eras/alonzo/test-suite/cddl-files/alonzo.cddl) is the [CDDL](https://tools.ietf.org/html/rfc8610) specification for after Plutus script support has been added.
+Cardano uses [CBOR](https://www.rfc-editor.org/info/rfc7049) as a serialization format. [Here](https://github.com/input-output-hk/cardano-ledger/blob/1cbf1fc2bb005a8206e5b5a7cdf44d35baaca455/eras/babbage/test-suite/cddl-files/babbage.cddl) is the [CDDL](https://tools.ietf.org/html/rfc8610) specification for Babbage era.
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/README.md:325]: Good idea (especially because of the tagged embedded CBOR), I added a tx example with Plutus elements in https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/64eed36fc454b1098e16de8d055c92b387225abd.

> perhaps we can remove even the existing description and just mention that one of the CBOR tools can be used to view details

I prefer keeping the structure description that we have - it is richer than what CBOR tools can display (it contains the tx element names).

File: common/protob/messages-cardano.proto (hunk 5 above)
Comment (davidmisiak) [reply common/protob/messages-cardano.proto]: Thanks for reminder, renamed in https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/8c333516c61e2f0990dea73d0183f529b2d44998 and https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/64eed36fc454b1098e16de8d055c92b387225abd. (I used only `legacy` and `babbage` in function and constant names, I find `array_legacy` and `map_babbage` too wordy for this purpose.)

File: core/src/apps/cardano/sign_tx/signer.py (hunk 6 above)
Comment (davidmisiak) [reply core/src/apps/cardano/sign_tx/signer.py:628]: I agree that the current implementation contains some duplication, but I'm not sure whether introducing multiple if-then-else statements is much better:

```python
title = "Inline datum" if chunk_type == "inline-datum" else "Reference script"
```
```python
br_type="confirm_inline_datum"
if chunk_type == "inline-datum"
else "confirm_reference_script"
```
```python
message_type = (
    messages.CardanoTxInlineDatumChunk
    if chunk_type == "inline-datum"
    else messages.CardanoTxReferenceScriptChunk
)
```

Personally, I slightly prefer the original implementation (I find it similarly "ugly", but a bit easier to reason about), but if you conclude that the extracted version is better, I'll be okay with that as well. Alternatively, we could ask Trezor devs what they think when opening the PR.

File: common/protob/messages-cardano.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (gabrielKerekes) [reply common/protob/messages-cardano.proto:218]: I rather agree it's too far-fetched. Maybe we should have reused the inputs message previously but it doesn't matter now. I was just curious what you thought.

File: core/src/apps/cardano/README.md [hunk 12]
```diff
@@ -385,3 +395,10 @@ a900818258203b40265111d8bb3c3c608d95b3a0bf83461ace32d79336579a1939b3aad1c0b70001
   }
 }
 ```
+
+### Babbage-era transaction body example
+This Plutus transaction body contains elements introduced in the Babbage era (perhaps the most intersting elements with regards to the CBOR encoding are the inline datum and the reference script). You can use one of the CBOR inspection tools listed on top of this document to examine the transaction body structure.
```
Comment (gabrielKerekes) [inline core/src/apps/cardano/README.md:400]:
```suggestion
This Plutus transaction body contains elements introduced in the Babbage era (perhaps the most interesting elements with regards to the CBOR encoding are the inline datum and the reference script). You can use one of the CBOR inspection tools listed on top of this document to examine the transaction body structure.
```

File: core/src/apps/cardano/layout.py [hunk 13]
```diff
@@ -215,16 +215,15 @@ async def confirm_sending(
     ctx: wire.Context,
     ada_amount: int,
     to: str,
-    is_change_output: bool,
+    output_message: str,
```
[the remark this answers is not in the document]
Comment (gabrielKerekes) [reply core/src/apps/cardano/layout.py:218]: Hm, perhaps we can use `address` instead of `third-party`? But I don't really have an issue with `third-party` 🤔 As that's what we generally use to talk about outputs with addresses.

[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/layout.py:218]: > But I don't really have an issue with third-party :thinking: As that's what we generally use to talk about outputs with addresses.

Yes, but device-owned outputs in Plutus tx are not given by address bytes, but address parameters (like change outputs), they are not `third-party` in any sense.

But `address` sounds good to me :+1:

File: core/src/apps/cardano/sign_tx/signer.py [hunk 14]
```diff
@@ -749,6 +969,125 @@ def _validate_required_signer(
...
+    def _should_show_collateral_return_init(
+        self, output: messages.CardanoTxOutput
+    ) -> bool:
+        if self.msg.total_collateral is None:
+            return True
+
+        if output.address_parameters is not None:  # change output
+            if not should_show_credentials(output.address_parameters):
+                return False
+
```
[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/sign_tx/signer.py]: We use `_is_change_output` to `determine what message to show to the user when confirming sending` (as stated in its docstring) and we override it in the `PlutusSigner` (since there are no change outputs, only device-owned outputs), so using it in `_is_simple_change_output` could affect the displaying logic.

I don't see harm in keeping the current implementation, because the condition `output.address_parameters is not None` is really simple and `_is_change_output` is used for something else.

File: core/src/apps/cardano/layout.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (davidmisiak) [reply core/src/apps/cardano/layout.py]: Implemented in https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/34f2cf4611ec15eb3565b7096a5dd690e23c5413.

File: core/src/apps/cardano/README.md (hunk 12 above)
Comment (davidmisiak) [reply core/src/apps/cardano/README.md]: Thanks for catching this! https://github.com/vacuumlabs/trezor-firmware/pull/78/commits/34f2cf4611ec15eb3565b7096a5dd690e23c5413

**Review by gabrielKerekes — APPROVED** [review]

**Comment by davidmisiak** [comment]
Merged in https://github.com/trezor/trezor-firmware/pull/2367
